File: src/main/java/com/back/together02be/infra/kis/rest/entity/KisAccessToken.java
package com.back.together02be.infra.kis.rest.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor
@Table(name = "kis_access_token")
public class KisAccessToken {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // 실제 access token 값
    @Column(nullable = false, length = 2000)
    private String accessToken;

    // 보통 Bearer
    @Column(nullable = false, length = 50)
    private String tokenType;

    // 토큰 만료 시각
    @Column(nullable = false)
    private LocalDateTime expiresAt;

    public KisAccessToken(String accessToken, String tokenType, LocalDateTime expiresAt) {
        this.accessToken = accessToken;
        this.tokenType = tokenType;
        this.expiresAt = expiresAt;
    }

    public void update(String accessToken, String tokenType, LocalDateTime expiresAt) {
        this.accessToken = accessToken;
        this.tokenType = tokenType;
        this.expiresAt = expiresAt;
    }

    // 만료 직전 에러 방지를 위해 1분 여유를 둔다.
    public boolean isUsable() {
        return expiresAt.isAfter(LocalDateTime.now().plusMinutes(1));
    }
}

File: src/main/kotlin/com/back/together02be/infra/kis/rest/entity/KisAccessToken.kt
package com.back.together02be.infra.kis.rest.entity

import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
@Table(name = "kis_access_token")
class KisAccessToken(
    // 주 생성자에는 값을 받아오는 역할만 담당합니다. (var/val 제거)
    accessToken: String,
    tokenType: String,
    expiresAt: LocalDateTime
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
        protected set

    // 클래스 내부 바디에서 프로퍼티를 선언하고 protected set을 걸어줍니다.
    @Column(nullable = false, length = 2000)
    var accessToken: String = accessToken
        protected set

    @Column(nullable = false, length = 50)
    var tokenType: String = tokenType
        protected set

    @Column(nullable = false)
    var expiresAt: LocalDateTime = expiresAt
        protected set

    fun update(accessToken: String, tokenType: String, expiresAt: LocalDateTime) {
        this.accessToken = accessToken
        this.tokenType = tokenType
        this.expiresAt = expiresAt
    }

    @get:Transient
    val isUsable: Boolean
        get() = expiresAt.isAfter(LocalDateTime.now().plusMinutes(1))
}